### PR TITLE
Use buffered writer when writing ExplodedJar to file

### LINKED ISSUE #823: OutOfMemoryError in ExplodeJarTask
**Build scan link**
Private, can't share unfortunately.

**Plugin version**
`v1.13.1`, however no changes in `ExplodeJarTask` in ` main..v1.13.1`

**Gradle version**
`7.4.2`

**(Optional) Android Gradle Plugin (AGP) version**
Not applicable.

**`reason` output for bugs relating to incorrect advice**
Not applicable.

**Describe the bug**
When serializing the output of `ExplodeJarTask` to file, it sometimes throws an `OutOfMemoryError`:
```
Caused by: java.lang.OutOfMemoryError: Java heap space
at okio.Buffer.readString(Buffer.kt:313)
at okio.Buffer.readUtf8(Buffer.kt:297)
at com.squareup.moshi.JsonAdapter.toJson(JsonAdapter.java:94)
at com.autonomousapps.tasks.ExplodeJarTask$ExplodeJarWorkAction.execute(ExplodeJarTask.kt:126)
```

Specifically the call `explodedJars.toJson()` causes this. To give some context, `explodedJars.toJson().length = 5385952` in our case. The daemon is configured to run with 8Gb already.

**To Reproduce**
It's not happening deterministically unfortunately,  but having an insane number of dependencies sometimes triggers this.

**Expected behavior**
No OOM exception.

**Additional context**
From some initial investigation the entire `Set<ExplodedJar>` is first serialized to json and then written to file. Maybe it makes more sense to serialize the `ExplodedJar` one by one?

### PR DESCRIPTION
**Pull request by daanschipper**
Fixes #823

### CODE REVIEW

**Comment by daanschipper** [comment]
Disclaimer: I've not used moshi before, open for suggestions on some improvement. I'm not sure about the pretty output, I'd was looking for something along the lines of `getJsonSetAdapter<ExplodedJar>().toPrettyJson()` but couldn't find it.

**Review by autonomousapps — APPROVED** [review]
Thank you, I really appreciate this! Will merge once green on CI.

**Comment by autonomousapps** [comment]
There are an awful lot of test failures, and unfortunately that means the build scan is too large to be viewed.

File: src/main/kotlin/com/autonomousapps/internal/utils/moshi.kt [hunk 1]
```diff
@@ -106,6 +108,32 @@ inline fun <reified K, reified V> Map<K, V>.toPrettyString(withNulls: Boolean =
...
+ */
+inline fun <reified T> File.bufferWriteJsonSet(jsonObject: Set<T>, indent: String = "") {
+  val writer = JsonWriter.of(this.sink().buffer())
+  getJsonSetAdapter<T>().indent(indent).toJson(writer, jsonObject)
+  writer.close() // need to explicitly close the writer
+}
+
+/**
+ * Buffers writes the json object to file using the indent to make the output human-readable.
+ * By default the output is compacted.
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/internal/utils/moshi.kt:126]: > Buffers writes of the set to disk, using the indent to make the output human-readable. By default, the output is compacted.

use that for both new methods :)

File: src/functionalTest/groovy/com/autonomousapps/jvm/ExplodeJarSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+import com.autonomousapps.jvm.projects.ExplodeJarProject
+import com.autonomousapps.kit.utils.Files
+
+import java.util.stream.Collectors
+
+import static com.autonomousapps.utils.Runner.build
+import static com.google.common.truth.Truth.assertThat
+
+final class ExplodeJarSpec extends AbstractJvmSpec {
```
Comment (autonomousapps) [inline src/functionalTest/groovy/com/autonomousapps/jvm/ExplodeJarSpec.groovy:11]: I'm a little worried about the fragility of this test. Any change to the relevant models will break the test, but that, to me, is just noise—not useful information.

I think you should just delete this spec, the `ExplodeJarProject`, and the two new resources files as well. I don't think they add anything useful when compared to the maintenance cost.

**Review by autonomousapps — CHANGES_REQUESTED** [review]
Thanks for the fix. Left some comments.

File: src/main/kotlin/com/autonomousapps/internal/utils/moshi.kt [hunk 3]
```diff
@@ -106,6 +108,32 @@ inline fun <reified K, reified V> Map<K, V>.toPrettyString(withNulls: Boolean =
...
+ *
+ * @param jsonObject The set to write to file
+ * @param indent The indent to control how the result is formatted
+ */
+inline fun <reified T> File.bufferWriteJson(jsonObject: T, indent: String = "") {
+  val writer = JsonWriter.of(this.sink().buffer())
+  getJsonAdapter<T>().indent(indent).toJson(writer, jsonObject)
+  writer.close() // need to explicitly close the writer
+}
```
[the remark this answers is not in the document]
Comment (daanschipper) [reply src/main/kotlin/com/autonomousapps/internal/utils/moshi.kt:135]: Was not aware of `.use { }`, that looks cleaner indeed!

**Review by autonomousapps — APPROVED** [review]
Thanks!

**Pull request merged**